File: app/services/tingwu_client.py
from __future__ import annotations

import time
import uuid
from dataclasses import dataclass
from datetime import datetime
from typing import Any

from app.config import get_settings


@dataclass
class TingwuTask:
    task_id: str
    task_status: str
    meeting_join_url: str
    started_at: float
    stopped_at: float | None = None


class TingwuClient:
    _mock_tasks: dict[str, TingwuTask] = {}

    def __init__(self) -> None:
        self.settings = get_settings()
        self.mode = self.settings.tingwu_mode.lower()
        self._sdk_client = None
        self._sdk_models = None

        if self.mode in {"sdk", "real"}:
            self._init_sdk_client()

    def _init_sdk_client(self) -> None:
        required = [
            self.settings.tingwu_access_key_id,
            self.settings.tingwu_access_key_secret,
            self.settings.tingwu_app_key,
        ]
        if not all(required):
            raise RuntimeError("Missing Tingwu credentials. Set TINGWU_ACCESS_KEY_ID/SECRET and TINGWU_APP_KEY")

        from alibabacloud_tea_openapi import models as open_api_models
        from alibabacloud_tingwu20230930 import models as tingwu_models
        from alibabacloud_tingwu20230930.client import Client as TingwuSdkClient

        config = open_api_models.Config(
            access_key_id=self.settings.tingwu_access_key_id,
            access_key_secret=self.settings.tingwu_access_key_secret,
            region_id=self.settings.tingwu_region,
        )
        config.endpoint = (
            self.settings.tingwu_endpoint
            if self.settings.tingwu_endpoint
            else f"tingwu.{self.settings.tingwu_region}.aliyuncs.com"
        )

        self._sdk_client = TingwuSdkClient(config)
        self._sdk_models = tingwu_models

    def create_realtime_task(self, task_key: str, title: str, webhook_url: str) -> dict[str, Any]:
        if self.mode == "mock":
            task_id = f"mock-{uuid.uuid4()}"
            join_url = f"wss://mock-tingwu.example.com/ws/{task_id}"
            self._mock_tasks[task_id] = TingwuTask(
                task_id=task_id,
                task_status="INITIATED",
                meeting_join_url=join_url,
                started_at=time.time(),
            )
            return {"task_id": task_id, "meeting_join_url": join_url}

        req = self._sdk_models.CreateTaskRequest(
            app_key=self.settings.tingwu_app_key,
            type="realtime",
            operation="start",
            input=self._sdk_models.CreateTaskRequestInput(
                task_key=task_key,
                source_language="cn",
                format="pcm",
                sample_rate=16000,
                progressive_callbacks_enabled=True,
            ),
            parameters=self._sdk_models.CreateTaskRequestParameters(
                meeting_assistance_enabled=True,
                meeting_assistance=self._sdk_models.CreateTaskRequestParametersMeetingAssistance(
                    types=["Actions", "KeyInformation"]
                ),
                summarization_enabled=True,
                summarization=self._sdk_models.CreateTaskRequestParametersSummarization(
                    types=["Paragraph", "Conversational", "QuestionsAnswering"]
                ),
                auto_chapters_enabled=True,
                transcription=self._sdk_models.CreateTaskRequestParametersTranscription(
                    diarization_enabled=True,
                    realtime_diarization_enabled=True,
                ),
            ),
        )

        response = self._sdk_client.create_task(req)
        data = (response.body.data.to_map() if response and response.body and response.body.data else {})

        task_id = data.get("TaskId")
        join_url = data.get("MeetingJoinUrl")
        if not task_id:
            raise RuntimeError(f"CreateTask missing TaskId: {response.body.to_map() if response.body else {}}")

        return {
            "task_id": task_id,
            "meeting_join_url": join_url,
            "raw": response.body.to_map() if response and response.body else {},
        }

    def stop_task(self, task_id: str) -> dict[str, Any]:
        if self.mode == "mock":
            task = self._mock_tasks.get(task_id)
            if not task:
                return {"ok": False, "error": "task_not_found"}
            task.task_status = "ONGOING"
            task.stopped_at = time.time()
            return {"ok": True}

        req = self._sdk_models.CreateTaskRequest(
            app_key=self.settings.tingwu_app_key,
            type="realtime",
            operation="stop",
            input=self._sdk_models.CreateTaskRequestInput(task_id=task_id, source_language="cn"),
        )
        response = self._sdk_client.create_task(req)
        body = response.body.to_map() if response and response.body else {}
        code = body.get("Code")
        if code and code != "0":
            return {"ok": False, "error": body.get("Message", "unknown_error"), "raw": body}
        return {"ok": True, "raw": body}

    def get_task_info(self, task_id: str) -> dict[str, Any]:
        if self.mode == "mock":
            task = self._mock_tasks.get(task_id)
            if not task:
                return {"TaskId": task_id, "TaskStatus": "FAILED", "ErrorMessage": "task_not_found"}

            if task.task_status == "ONGOING" and task.stopped_at and time.time() - task.stopped_at > 2:
                task.task_status = "COMPLETED"

            payload: dict[str, Any] = {
                "TaskId": task.task_id,
                "TaskStatus": task.task_status,
                "MeetingJoinUrl": task.meeting_join_url,
            }
            if task.task_status == "COMPLETED":
                payload["InlineResult"] = {
                    "transcript": [
                        {
                            "speaker": "说话人1",
                            "start_ms": 0,
                            "end_ms": 1200,
                            "text": "这是一次关于AI听会后端方案的讨论。",
                        },
                        {
                            "speaker": "说话人2",
                            "start_ms": 1200,
                            "end_ms": 3200,
                            "text": "我们决定采用通义听悟直连并在完成态入库。",
                        },
                    ],
                    "summary": {
                        "overview": "会议明确采用听悟直连、完成态入库与无MQ异步任务表方案。",
                        "chapters": [
                            {"title": "架构决策", "start": "00:00", "content": "确认技术路径"}
                        ],
                        "decisions": ["前端直连听悟WS", "后端回调+轮询兜底", "不使用RocketMQ"],
                        "highlights": ["完成态入库"],
                        "todos": [
                            {
                                "text": "完善生产环境TINGWU SDK实现",
                                "owner": "研发",
                            }
                        ],
                    },
                }
            return payload

        response = self._sdk_client.get_task_info(task_id)
        body_map = response.body.to_map() if response and response.body else {}
        data = body_map.get("Data", {}) or {}
        result = data.get("Result", {}) or {}

        payload: dict[str, Any] = {
            "TaskId": data.get("TaskId") or task_id,
            "TaskStatus": data.get("TaskStatus") or body_map.get("Code"),
            "Result": result,
            "OutputMp3Path": data.get("OutputMp3Path"),
            "OutputMp4Path": data.get("OutputMp4Path"),
            "RawResponse": body_map,
        }
        if data.get("ErrorCode"):
            payload["ErrorCode"] = data.get("ErrorCode")
        if data.get("ErrorMessage"):
            payload["ErrorMessage"] = data.get("ErrorMessage")
        return payload

    @staticmethod
    def extract_task_id(payload: dict[str, Any]) -> str | None:
        for key in ("TaskId", "taskId", "task_id"):
            value = payload.get(key)
            if isinstance(value, str) and value:
                return value
        data = payload.get("Data")
        if isinstance(data, dict):
            return TingwuClient.extract_task_id(data)
        return None

    @staticmethod
    def extract_task_status(payload: dict[str, Any]) -> str | None:
        for key in ("TaskStatus", "taskStatus", "task_status"):
            value = payload.get(key)
            if isinstance(value, str) and value:
                return value.upper()
        data = payload.get("Data")
        if isinstance(data, dict):
            return TingwuClient.extract_task_status(data)
        return None

    @staticmethod
    def extract_result_urls(task_info: dict[str, Any]) -> dict[str, str]:
        candidates: dict[str, str] = {}

        result = task_info.get("Result")
        if isinstance(result, dict):
            if isinstance(result.get("Transcription"), str):
                candidates["transcript"] = result["Transcription"]
            if isinstance(result.get("Summarization"), str):
                candidates["summary"] = result["Summarization"]
            if isinstance(result.get("AutoChapters"), str):
                candidates["chapters"] = result["AutoChapters"]
            if isinstance(result.get("MeetingAssistance"), str):
                candidates["meeting_assistance"] = result["MeetingAssistance"]

        for media_key, media_name in (("OutputMp3Path", "mp3"), ("OutputMp4Path", "mp4")):
            media_url = task_info.get(media_key)
            if isinstance(media_url, str) and media_url.startswith("http"):
                candidates[media_name] = media_url

        def walk(node: Any, path: str = "") -> None:
            if isinstance(node, dict):
                for k, v in node.items():
                    lower_key = k.lower()
                    current = f"{path}.{lower_key}" if path else lower_key
                    if isinstance(v, str) and v.startswith("http"):
                        if "summary" in lower_key or "summary" in current:
                            candidates.setdefault("summary", v)
                        elif "chapter" in lower_key:
                            candidates.setdefault("chapters", v)
                        elif "meetingassistance" in lower_key or "assist" in lower_key:
                            candidates.setdefault("meeting_assistance", v)
                        elif "trans" in lower_key or "text" in lower_key:
                            candidates.setdefault("transcript", v)
                        else:
                            candidates.setdefault("raw", v)
                    else:
                        walk(v, current)
            elif isinstance(node, list):
                for item in node:
                    walk(item, path)

        walk(task_info)
        return candidates


class TaskEventNormalizer:
    @staticmethod
    def normalize(payload: dict[str, Any]) -> dict[str, Any]:
        client = TingwuClient()
        return {
            "task_id": client.extract_task_id(payload),
            "task_status": client.extract_task_status(payload),
            "event_time": payload.get("EventTime") or datetime.utcnow().isoformat(),
            "raw": payload,
        }

File: src/main/java/com/gczm/aimeeting/service/TingwuClient.java
package com.gczm.aimeeting.service;

import com.aliyun.auth.credentials.Credential;
import com.aliyun.auth.credentials.provider.ICredentialProvider;
import com.aliyun.auth.credentials.provider.StaticCredentialProvider;
import com.aliyun.sdk.service.tingwu20230930.AsyncClient;
import com.aliyun.sdk.service.tingwu20230930.models.CreateTaskRequest;
import com.aliyun.sdk.service.tingwu20230930.models.CreateTaskResponse;
import com.aliyun.sdk.service.tingwu20230930.models.GetTaskInfoRequest;
import com.aliyun.sdk.service.tingwu20230930.models.GetTaskInfoResponse;
import com.gczm.aimeeting.common.ApiException;
import com.gczm.aimeeting.config.AppProperties;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import darabonba.core.client.ClientOverrideConfiguration;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletionException;

@Service
@RequiredArgsConstructor
public class TingwuClient {

    private final AppProperties appProperties;

    private AsyncClient cachedClient;

    public Map<String, Object> createRealtimeTask(String taskKey, String title, String webhookUrl) {
        if (!"sdk".equalsIgnoreCase(appProperties.getTingwu().getMode())) {
            throw new ApiException(HttpStatus.INTERNAL_SERVER_ERROR, "TINGWU_MODE must be sdk");
        }

        CreateTaskRequest.Input input = CreateTaskRequest.Input.builder()
                .taskKey(taskKey)
                .sourceLanguage("cn")
                .format("pcm")
                .sampleRate(16000)
                .progressiveCallbacksEnabled(true)
                .build();

        CreateTaskRequest.MeetingAssistance meetingAssistance = CreateTaskRequest.MeetingAssistance.builder()
                .types(List.of("Actions", "KeyInformation"))
                .build();

        CreateTaskRequest.Summarization summarization = CreateTaskRequest.Summarization.builder()
                .types(List.of("Paragraph", "Conversational", "QuestionsAnswering"))
                .build();

        CreateTaskRequest.Transcription transcription = CreateTaskRequest.Transcription.builder()
                .diarizationEnabled(true)
                .realtimeDiarizationEnabled(true)
                .build();

        CreateTaskRequest.Parameters parameters = CreateTaskRequest.Parameters.builder()
                .meetingAssistanceEnabled(true)
                .meetingAssistance(meetingAssistance)
                .summarizationEnabled(true)
                .summarization(summarization)
                .autoChaptersEnabled(true)
                .transcription(transcription)
                .build();

        CreateTaskRequest request = CreateTaskRequest.builder()
                .appKey(appProperties.getTingwu().getAppKey())
                .type("realtime")
                .operation("start")
                .input(input)
                .parameters(parameters)
                .build();

        try {
            CreateTaskResponse response = client().createTask(request).join();
            Map<String, Object> raw = response.getBody() == null ? Map.of() : response.getBody().toMap();
            String taskId = response.getBody() == null || response.getBody().getData() == null
                    ? null : response.getBody().getData().getTaskId();
            String meetingJoinUrl = response.getBody() == null || response.getBody().getData() == null
                    ? null : response.getBody().getData().getMeetingJoinUrl();

            if (StringUtils.isBlank(taskId)) {
                throw new ApiException(HttpStatus.BAD_GATEWAY, "CreateTask missing TaskId");
            }

            Map<String, Object> result = new HashMap<>();
            result.put("task_id", taskId);
            result.put("meeting_join_url", meetingJoinUrl);
            result.put("raw", raw);
            return result;
        } catch (CompletionException ex) {
            Throwable root = ex.getCause() == null ? ex : ex.getCause();
            throw new ApiException(HttpStatus.BAD_GATEWAY, "Tingwu request failed: " + root.getMessage());
        } catch (Exception ex) {
            throw new ApiException(HttpStatus.BAD_GATEWAY, "Tingwu request failed: " + ex.getMessage());
        }
    }

    public Map<String, Object> stopTask(String taskId) {
        CreateTaskRequest.Input input = CreateTaskRequest.Input.builder()
                .taskId(taskId)
                .sourceLanguage("cn")
                .build();

        CreateTaskRequest request = CreateTaskRequest.builder()
                .appKey(appProperties.getTingwu().getAppKey())
                .type("realtime")
                .operation("stop")
                .input(input)
                .build();

        try {
            CreateTaskResponse response = client().createTask(request).join();
            String code = response.getBody() == null ? null : response.getBody().getCode();
            String message = response.getBody() == null ? null : response.getBody().getMessage();

            Map<String, Object> result = new HashMap<>();
            result.put("ok", StringUtils.isBlank(code) || "0".equals(code));
            result.put("error", message == null ? "" : message);
            result.put("raw", response.getBody() == null ? Map.of() : response.getBody().toMap());
            return result;
        } catch (CompletionException ex) {
            Throwable root = ex.getCause() == null ? ex : ex.getCause();
            throw new ApiException(HttpStatus.BAD_GATEWAY, "Tingwu request failed: " + root.getMessage());
        } catch (Exception ex) {
            throw new ApiException(HttpStatus.BAD_GATEWAY, "Tingwu request failed: " + ex.getMessage());
        }
    }

    public Map<String, Object> getTaskInfo(String taskId) {
        GetTaskInfoRequest request = GetTaskInfoRequest.builder().taskId(taskId).build();

        try {
            GetTaskInfoResponse response = client().getTaskInfo(request).join();
            var body = response.getBody();
            var data = body == null ? null : body.getData();
            var result = data == null ? null : data.getResult();

            Map<String, Object> payload = new HashMap<>();
            payload.put("TaskId", data == null ? taskId : data.getTaskId());
            payload.put("TaskStatus", data == null ? null : data.getTaskStatus());
            payload.put("Result", result == null ? Map.of() : result.toMap());
            payload.put("OutputMp3Path", data == null ? null : data.getOutputMp3Path());
            payload.put("OutputMp4Path", data == null ? null : data.getOutputMp4Path());
            payload.put("RawResponse", body == null ? Map.of() : body.toMap());
            if (data != null && data.getErrorCode() != null) {
                payload.put("ErrorCode", data.getErrorCode());
            }
            if (data != null && data.getErrorMessage() != null) {
                payload.put("ErrorMessage", data.getErrorMessage());
            }
            return payload;
        } catch (CompletionException ex) {
            Throwable root = ex.getCause() == null ? ex : ex.getCause();
            throw new ApiException(HttpStatus.BAD_GATEWAY, "Tingwu request failed: " + root.getMessage());
        } catch (Exception ex) {
            throw new ApiException(HttpStatus.BAD_GATEWAY, "Tingwu request failed: " + ex.getMessage());
        }
    }

    public String extractTaskStatus(Map<String, Object> payload) {
        String value = findString(payload, List.of("TaskStatus", "taskStatus", "task_status"));
        return value == null ? null : value.toUpperCase();
    }

    public String extractTaskId(Map<String, Object> payload) {
        return findString(payload, List.of("TaskId", "taskId", "task_id"));
    }

    public Map<String, String> extractResultUrls(Map<String, Object> taskInfo) {
        Map<String, String> candidates = new HashMap<>();
        Map<String, Object> result = asMap(taskInfo.get("Result"));

        addIfHttp(candidates, "transcript", result.get("Transcription"));
        addIfHttp(candidates, "summary", result.get("Summarization"));
        addIfHttp(candidates, "chapters", result.get("AutoChapters"));
        addIfHttp(candidates, "meeting_assistance", result.get("MeetingAssistance"));

        addIfHttp(candidates, "mp3", taskInfo.get("OutputMp3Path"));
        addIfHttp(candidates, "mp4", taskInfo.get("OutputMp4Path"));

        walkForUrls(taskInfo, "", candidates);
        return candidates;
    }

    public Map<String, Object> normalizeEvent(Map<String, Object> payload) {
        Map<String, Object> normalized = new HashMap<>();
        normalized.put("task_id", extractTaskId(payload));
        normalized.put("task_status", extractTaskStatus(payload));
        normalized.put("event_time", payload.getOrDefault("EventTime", LocalDateTime.now().toString()));
        normalized.put("raw", payload);
        return normalized;
    }

    private synchronized AsyncClient client() {
        if (cachedClient != null) {
            return cachedClient;
        }

        if (StringUtils.isAnyBlank(
                appProperties.getTingwu().getAccessKeyId(),
                appProperties.getTingwu().getAccessKeySecret(),
                appProperties.getTingwu().getAppKey())) {
            throw new ApiException(HttpStatus.INTERNAL_SERVER_ERROR,
                    "Missing Tingwu credentials. Set TINGWU_ACCESS_KEY_ID/SECRET and TINGWU_APP_KEY");
        }

        Credential credential = new Credential.Builder()
                .accessKeyId(appProperties.getTingwu().getAccessKeyId())
                .accessKeySecret(appProperties.getTingwu().getAccessKeySecret())
                .build();
        ICredentialProvider provider = StaticCredentialProvider.create(credential);

        ClientOverrideConfiguration overrideConfiguration = ClientOverrideConfiguration.create()
                .setProtocol("https");
        if (StringUtils.isNotBlank(appProperties.getTingwu().getEndpoint())) {
            overrideConfiguration.setEndpointOverride(appProperties.getTingwu().getEndpoint());
        }

        cachedClient = AsyncClient.builder()
                .credentialsProvider(provider)
                .region(appProperties.getTingwu().getRegion())
                .overrideConfiguration(overrideConfiguration)
                .build();
        return cachedClient;
    }

    private void walkForUrls(Object node, String path, Map<String, String> candidates) {
        if (node instanceof Map<?, ?> map) {
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                if (!(entry.getKey() instanceof String key)) {
                    continue;
                }
                String lowerKey = key.toLowerCase();
                String current = path.isBlank() ? lowerKey : path + "." + lowerKey;
                Object value = entry.getValue();
                if (value instanceof String s && s.startsWith("http")) {
                    if (lowerKey.contains("summary") || current.contains("summary")) {
                        candidates.putIfAbsent("summary", s);
                    } else if (lowerKey.contains("chapter")) {
                        candidates.putIfAbsent("chapters", s);
                    } else if (lowerKey.contains("meetingassistance") || lowerKey.contains("assist")) {
                        candidates.putIfAbsent("meeting_assistance", s);
                    } else if (lowerKey.contains("trans") || lowerKey.contains("text")) {
                        candidates.putIfAbsent("transcript", s);
                    } else {
                        candidates.putIfAbsent("raw", s);
                    }
                } else {
                    walkForUrls(value, current, candidates);
                }
            }
            return;
        }

        if (node instanceof List<?> list) {
            for (Object item : list) {
                walkForUrls(item, path, candidates);
            }
        }
    }

    private String findString(Map<String, Object> payload, List<String> keys) {
        for (String key : keys) {
            Object value = payload.get(key);
            if (value instanceof String s && StringUtils.isNotBlank(s)) {
                return s;
            }
        }
        Object data = payload.get("Data");
        if (data instanceof Map<?, ?> map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> nested = (Map<String, Object>) map;
            return findString(nested, keys);
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> asMap(Object value) {
        if (value instanceof Map<?, ?> map) {
            return (Map<String, Object>) map;
        }
        return new HashMap<>();
    }

    private void addIfHttp(Map<String, String> out, String key, Object value) {
        if (value instanceof String s && s.startsWith("http")) {
            out.put(key, s);
        }
    }
}
